### ISSUE TITLE
Run the tool to correct a corrupted backup file

### ISSUE THREAD

**Issue by neo1992** [bepaald/signalbackup-tools#62]
Coming here based on your recommendation here: https://github.com/signalapp/Signal-Android/issues/12057

So what I want to achieve is to correct error in my backup file and try to import the correct backup.

I'm running the command like so: `signalbackup-tools_win.exe signal-2022-03-11-14-04-18.backup 946304659236220026494649192649`

Here's the output after running it for a few hours, it's 11GB in total so not sure how long this will take.

Via pastebin as it's 2x longer than allowed here: https://pastebin.com/MgdXGxD1

Will this achieve my goal or am I missing something?

**Comment by bepaald** [comment]
Thanks. I don't know if it is still running, but you can surely cancel it now, I have enough to look at.

At the moment, it looks most likely there is no valid data in the backup after the corruption, but I do see something strange in the output that I need to investigate when I have time (hopefully tomorrow).

**Comment by bepaald** [comment]
There was actually a bit of a bug in the recovery code, that appears only for very large files (such as yours). I think I have just fixed it, so if you want to run it again (with the new version) that might be helpful. No need to let it run for hours, maybe just a couple of minutes after the first error (at around 50%) should be more than enough to see what's going on.

(It's way past my bed time here, so I'll get back to it tomorrow)

Thanks.

**Comment by neo1992** [comment]
@bepaald thank you so much for going this extra mile! I ran your latest version and here's the output after running it for ~30mins, not sure if this is useful to you: 

C:\test\signalbackup-tools_win.exe signal-2022-03-11-14-04-18.backup 946304659236220026494649192649
signalbackup-tools (signalbackup-tools_win.exe) source version 20220314.230546 (OpenSSL)
IV: (hex:) <removed> (size: 16)
SALT: (hex:) <removed> (size: 32)
BACKUPKEY: (hex:) <removed> (size: 32)
CIPHERKEY: (hex:) <removed> (size: 32)
MACKEY: (hex:) <removed> (size: 32)
COUNTER: 3373360406
Reading backup file...
FRAME 10010 (050.5%)... Failed to read next frame (0 bytes at filepos 6128448903)
Starting bruteforcing offset to next valid frame... starting after: 6128448903
Checking offset 95008530 bytes

**Comment by bepaald** [comment]
Ok, I think that's it then, as I suspected the only table that is (partially) present is the _'part'_ table, which contains the message attachments. I think the current methods of saving these require a full working backup, but I could easily write something that will save any attachments that are found in an incomplete backup such as yours. Let me know if you're interested and I'll probably get it done in a day or two.

**Comment by neo1992** [comment]
Just so I understand, I can only restore attachments and not the messages themselves? In any case, if you could help me out that'd be highly appreciated!!

**Comment by bepaald** [comment]
>Just so I understand, I can only restore attachments and not the messages themselves?

Yes this is correct, I'm sorry. The tables with the messages are simply not there anymore. It would have been great if they were in the beginning of the file, I even wrote a forum message about it back when the order of the tables was changed (https://community.signalusers.org/t/change-order-of-data-in-backup-file-to-make-it-more-resilient-to-errors/35795), but it got little attention. I'm not sure how a backup ends up in the state yours is in, but recently I've seen similar cases where the storage was actually simply full. Is there any chance there wasn't enough space for the backup to be written on the phone (it was quite big)?

Anyway, I made a little start of the function to extract the media, so it'll be done tomorrow no problem. Normally, I export the media all sorted by which contact sent it, and with correct timestamps and such, but all this info is missing as well, so it'll just be one big directory with everything in it and only approximate timestamps. I'll let you know tomorrow when it's ready with some instructions.
